### ISSUE TITLE
Support for RKE2 secrets encryption keys rotation

### ISSUE THREAD

**Issue by snasovich** [rancher/rancher#35436]
Implement support to rotate secrets encryption keys. This will  allow the UI equivalent to the following RKE1 UI to be implemented:

Rotate encryption keys:
![image](https://user-images.githubusercontent.com/85187633/141045220-b2f8b75e-ad81-4804-9c3f-23712ec5b51a.png)

This requires https://github.com/rancher/rke2/issues/748 implemented on RKE2 side.

**Comment by jakefhyde** [comment]
## How to rotate encryption keys

Although there will be a UI button to start rotating encryption keys before the `v2.6.4` release, currently the encryption keys can be rotated by adding the following:
```
rotateEncryptionKeys:
  generation: 1
```
under the `rkeConfig` section of a cluster object's yaml.

## Additional test cases

- single node rke2/k3s
- 3 node control plane rke2/k3s
- creating the cluster with `rotateEncryptionKeys.generation` in the spec (cluster should not rotate while provisioning).
- \>5 node control plane

**Comment by timhaneunsoo** [comment]
### Test Environment: ###
Rancher version: v2.6-head
Rancher cluster type: HA
Docker version: 20.10

Downstream cluster type: RKE2 node driver and K3s node driver

----------

### Testing: ###

Tested the following:

Single Node rke2/k3s
1. Create new cluster with encryption enabled (rke2/k3s)
2. Create new cluster and edit cluster and update it and enable encryption (k3s)
3. Disable/Enable encryption (k3s)
4. Add Remove control plane (rke2/k3s)
5. Create a new cluster with custom encryption configuration (rke2/k3s)
6. Create new cluster and edit cluster and update it and enable custom encryption (k3s)
7. Disable encryption (custom encryption) (k3s)
8. Add/Remove control planes (custom encryption) (rke2/k3s)
9. Edit custom encryption (rke2/k3s)
10. Switch from default encryption to custom encryption (rke2/k3s)

The same testing for 3 node control plane rke2/k3s and also for >5 node control plane

Create cluster with `rotateEncryuptionKeys.generation` in the spec

**Result** - Fail

All scenarios failed, spoke with @jakefhyde for all necessary logs and information

Create cluster with `rotateEncryuptionKeys.generation` in the spec - causes worker nodes to be stuck in `Provisioning` state for rke2 and causes etcd nodes to be stuck in `Provisioning` for k3s nodes.

**Comment by snasovich** [comment]
Ths is blocked by https://github.com/rancher/rke2/issues/2562 at this time.

**Comment by snasovich** [comment]
Moving to 2.6.5, would like to hide the action on the UI. @Sahota1225 to open dashboard ticket and link it here.

**Comment by Sahota1225** [comment]
https://github.com/rancher/dashboard/issues/5379

**Comment by gaktive** [comment]
Thanks @snasovich . So for 2.6.4, is UI doing anything for https://github.com/rancher/dashboard/issues/4540 if https://github.com/rancher/dashboard/issues/5379 is the subsequent fix?

**Comment by snasovich** [comment]
@gaktive , left a clarifying comment on the other issue: https://github.com/rancher/dashboard/issues/4540#issuecomment-1065796497

**Comment by snasovich** [comment]
This is now ready to test on v2.6-head. The testing note from https://github.com/rancher/rancher/issues/35436#issuecomment-1048918154 still applies until the UI part (https://github.com/rancher/dashboard/issues/4540) is ready.

**Comment by timhaneunsoo** [comment]
Update: issues were found where the secrets encryption rotation is getting stuck during the process. Discussed with @Oats87 and this ticket is being reopened.

**Comment by snasovich** [comment]
Due to blocking issues that are yet to be resolved and the general agreement we're likely to run into more issues with testing, the current plan is to not include Encryption Key Rotation functionality into 2.6.5 release. We're going to leave `rotateEncryptionKeys.generation` field in, but bumping it won't do anything in 2.6.5.

When announcing RKE2 Provisioning GA on Rancher release notes, we will need to make sure to have a section like "Limitations and Known Issues" where no support for Encryption Key Rotation functionality will need to be called out explicitly.

Leaving milestone as 2.6.5 for inclusion on release notes and will bump milestone to 2.6.6 after it's taken care of.

**Comment by jakefhyde** [comment]
Blocked waiting for rke2 and k3s kdm june releases since encryption key rotation is feature gated in KDM.

**Comment by thaneunsoo** [comment]
### Test Environment: ###
Rancher version: v2.6-head 2c21373
Rancher cluster type: HA
Docker version: 20.10

----------

### Testing: ###

Tested all scenarios listed [here](https://confluence.suse.com/display/RANQA/RKE2+Encryption+Key+Rotation+Test+Plan)

**Result**
All scenarios have passed besides rotating large number of secrets.
Issues found are:
- [38277](https://github.com/rancher/rancher/issues/38277)
- [38283](https://github.com/rancher/rancher/issues/38283)

Closing this ticket as testing is complete and issues will be tracked in the tickets listed above.
